File: it-agents/skills/code-architect/references/code-examples/java/Prototype.java
// Prototype — copy existing objects via a clone interface without coupling to classes.

import java.util.ArrayList;
import java.util.List;

interface Copyable<T> {
    T copy();
}

class Shape implements Copyable<Shape> {
    int x, y;
    String color;
    Shape(int x, int y, String color) { this.x = x; this.y = y; this.color = color; }
    public Shape copy() { return new Shape(x, y, color); }
    public String toString() { return color + "@(" + x + "," + y + ")"; }
}

class Group implements Copyable<Group> {
    final List<Shape> shapes = new ArrayList<>();
    void add(Shape s) { shapes.add(s); }
    public Group copy() {
        Group g = new Group();
        for (Shape s : shapes) g.add(s.copy()); // deep copy of members
        return g;
    }
    public String toString() { return shapes.toString(); }
}

public class Prototype {
    public static void main(String[] args) {
        Group original = new Group();
        original.add(new Shape(1, 2, "red"));
        original.add(new Shape(3, 4, "blue"));

        Group clone = original.copy();
        clone.shapes.get(0).color = "green"; // mutate clone only

        System.out.println("original: " + original);
        System.out.println("clone:    " + clone);
    }
}

File: it-agents/skills/code-architect/references/code-examples/csharp/Prototype.cs
// Prototype — copy existing objects via a clone interface without coupling to their classes.
using System;
using System.Collections.Generic;

namespace Patterns.Prototype
{
    interface IPrototype<T>
    {
        T Clone();
    }

    sealed class ConcretePrototype : IPrototype<ConcretePrototype>
    {
        public string Label { get; set; }
        public List<int> Values { get; set; }

        public ConcretePrototype(string label, List<int> values)
        {
            Label = label;
            Values = values;
        }

        // Deep copy: mutable members are duplicated, not shared.
        public ConcretePrototype Clone() => new(Label, new List<int>(Values));

        public override string ToString() => $"{Label}:[{string.Join(",", Values)}]";
    }

    static class Demo
    {
        public static void Run()
        {
            var original = new ConcretePrototype("origin", new List<int> { 1, 2, 3 });
            ConcretePrototype copy = original.Clone();
            copy.Label = "copy";
            copy.Values.Add(4);

            Console.WriteLine(original); // unaffected by changes to copy
            Console.WriteLine(copy);
        }

        public static void Main() => Run();
    }
}
